rfc/rfc006-distributed-registry.md

```diff
@@ -0,0 +1,146 @@
+# RFC006 Distributed Registry
+
+|  |  |
+| :--- | :--- |
+| Nuts foundation | R.G. Krul |
+| Request for Comments: 006 | Nedap |
+|  | September 2020 |
+
+## Distributed Registry
+
+### Abstract
+
+This RFC describes a protocol to build a registry containing information required for care organizations to exchange data
+through their software vendors. The registry typically contains vendors, organizations, services and endpoints.
+It uses [RFC004](rfc004-distributed-document-format.md)'s documents as underlying data format.
+
+### Status of document
+
+This document is currently a draft.
+
+### Copyright Notice
+
+![](../.gitbook/assets/license.png)
+
+This document is released under the [Attribution-ShareAlike 4.0 International \(CC BY-SA 4.0\) license](https://creativecommons.org/licenses/by-sa/4.0/).
+
+## 1.  Introduction
+
+When care organizations want to exchange data using Nuts they need to know where to find that data and how to authenticate it.
+This RFC describes the document types containing the information required to form this registry.
+It uses [RFC004](rfc004-distributed-document-format.md) as underlying data format. The origin and requirements of
+certificates are defined by RFC Certificate Structure (TODO).
+
+## 2. Terminology
+
+* **Document**: a distributed document as specified by [RFC004](rfc004-distributed-document-format.md).
+* **Organization**: a care organization exchanging data with other care organizations over a Nuts network.
+* **Vendor**: an entity developing software for care organizations participating in a Nuts network. Organizations
+  access the Nuts network through their vendor's software.
+* **Endpoint**: typically a URI or URL exposed by a vendor which can be used by other vendors to pull data.
+* **Service**: a grouping of endpoints implementing a service required by a Bolt.
+
+Other terminology comes from the [RFC001 Nuts Start Architecture](rfc001-nuts-start-architecture.md#nuts-start-architecture).
+
+## 3. Updating
```

Comment:
Does an update make a previous document obsolete? How does this work with timelined data?

Reply:
Can you give an example?

Reply:
What happens when you change the validity period of an endpoint? Is the existing record updated with a new record that adds an end_date plus a new record that has a start_date starting at the previous end_date? That last record would then not be an update just a new record.

Reply:
In the current state (complete replacement) you'd indeed create a new record with the same ID and properties but with updated validity, which effectively replaces the endpoint. 

IMO updates through complete object replacement is the easiest way. The alternative is allowing incremental updates to JSON documents instead of replacement. In that case we should use a standard like http://jsonpatch.com/).

It all matters on the requirements of the system:

**Data size**
Do we care about bandwidth/storage? Incremental updates lead to less data.

**Simplicity**
How important do we find simplicity? Object replacement is simpler.

**Domain modelling**
Do we foresee somebody would want to fire domain-specific events when a particular property changes? This is easy with incremental updates but very hard/impossible with complete replacement because you'd have to derive the domain-specific event by checking deltas of object states. For example a change of `Endpoint.StartDate` can be mapped to an `EndpointValidityUpdatedEvent` but what if multiple properties change in the same delta? Fire 2 events but in what order?

(this is an actual problem I encountered in a system I built)

Reply:
Summary of discussion: separate distribution model of internal model. The distribution model can work with jsonpatch, but the internal model should store versions and when a specific version was valid.